conf-ida 0.3

gitoleg opened the pull request:
The main outcome is that we're no longer using the `find`/`mdfind` utilities to find the IDA Pro installation, meaning that this PR makes our search less time-consuming (fixes https://github.com/ocaml/opam-repository/issues/15643)

## The Search

Both for Linux and macOS we start searching for IDA 64-bit executables in `PATH` using `which`. Then, if it fails, we continue searching in the OS-specific directories. We search until the first success, no deeper than 4 subdirectories, and examine no more than 1024 paths. 

## The macOS details

On macOS, IDA Pro 6.x and below is installed into two folders, `idaq.app` and `idaq64.app`, and each of them containing all the binaries we need (at `Contents/MacOS`). After a quick search on the Internet, we figured out that `IDA-7` has changed the naming scheme to `ida.app` and `ida64.app`.

We examine the following folders, `/Applications/`, `$HOME/Applications`, and `$HOME` for the `IDA` for the `*.app` folders in the following order
1. `ida64`
2. `ida`
3. `idaq64`
4. `idaq`

This gives priority to the more modern and 64-bit versions.

## The Linux details
On Linux, we try to find IDA 64 bit executables in the `$HOME` folder.

## Customization

It is possible to disable searching and set the concrete path to the IDA Pro installation using either the `IDA_PATH` environment variable or the `ida-path` opam variable.

gitoleg committed:
limits time for find commands

gitoleg committed:
don't use find or mdfind

gitoleg committed:
find ida64 only on linux

gitoleg committed:
refactoring

ivg commented on packages/conf-ida/conf-ida.0.2/files/find-ida.ml.in:
```diff
@@ -56,81 +91,92 @@ let exn_to_string = function
-  match find_at_home () with
-  | None -> find_ida64 "/"
-  | x -> x
+  match List.find_opt ~f:contains_ida64 @@ all_pathes () with
```
it is `paths` not `pathes`

ivg replied:
You shouldn't be searching in the `PATH` constituents, you shall use `which` for that. Use linear search only for paths, that are not in `PATH`.

ivg reviewed (changes requested):
Looks good, but do not search in PATH, it is already searched by `which`. If you're also searching in PATH in the corresponding binutils patch, please update it also.

gitoleg committed:
don't search in the $PATH: we already use which

gitoleg committed:
fixes bug

gitoleg committed:
fixes bug

gitoleg committed:
bumps a new version

gitoleg committed:
few minor updates

gitoleg committed:
looks for ida prefix on macOS

gitoleg committed:
adds depth to the search

gitoleg committed:
applies depth regardless of path

gitoleg committed:
search for ida*64.app

gitoleg committed:
uses known names

gitoleg committed:
improved macOS search:

the order is as the following:
1. ida64.app
2. ida.app
3. idaq64.app
4. idaq.app

gitoleg committed:
refactored

gitoleg committed:
reimplemented the search

Now, a newer version will be chosen if it appears on the
same level as an older one.

gitoleg committed:
simplified the search

gitoleg committed:
updated description

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+
+let ida6 = "idaq"
+let ida7 = "ida"
+let idas = [ ida7; ida6; ]
```
there is no need to introduce extra variables, they only increase the cognition burden and, in fact are wrongly named, `ida6` is also `ida5`, `ida4`, and so on, and `ida7` could also become `ida8`, `ida9`, etc. 
```suggestion
let ida = [ "ida64"; "idaq64"]
let ida_apps = ["ida64.app"; "ida.app"]
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+    | None -> first_success fs
+    | x -> x
+
+let name64 x = x ^ "64"
```
an extra and totally unnecessary complexity
```suggestion
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+let which name () = cmd "which %s" name
+
+let which_ida64 () =
+  first_success @@ List.map idas ~f:(fun ida -> which (name64 ida))
```
```suggestion
  first_success @@ List.map idas ~f:which
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+let contains_ida64 path =
+  Array.exists (Sys.readdir path) ~f:(fun file ->
+      List.exists idas ~f:(fun ida ->
+          file = name64 ida))
```
```suggestion
      List.exists idas ~f:(List.mem file idas))
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+    Sys.readdir path |>
+    Array.fold_left ~init:[] ~f:(fun acc p ->
+        let path' = path / p in
+        if is_dir path' then path' :: acc
```
having `path'` instead of `path` presumes that `path` is used later. 
```suggestion
        let path = path / p in
        if is_dir path then path :: acc
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+end
+
+let max_depth = 4
+let total = 1024
```
```suggestion
let max_paths = 1024
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+
+let find path test =
+  let rec loop depth entries path =
+    if depth = max_depth || entries >= total then None
```
```suggestion
    if depth > max_depth || entries > max_paths then None
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+  let rec loop depth entries path =
+    if depth = max_depth || entries >= total then None
+    else
+      let dirs = folders path in
```
A folder may easily contain thousands of entries, remove the `folders` function and check if the path is a directory inside the `loop` function. You should also count _all_ entries, including folders and files, to the total maximum number of paths. 

E.g., you can easily hit a folder that has a million of entries, a directory that has an infinite number of paths (e.g., when they are generated on the fly).

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+      match List.find_opt dirs ~f:test with
+      | None ->
+        List.find_map dirs
+          ~f:(loop (depth + 1) (entries + List.length dirs))
```
again, there could be millions of entries in dirs, so calling `List.length` should be avoided, especially since you can easily incorporate the counting of paths inside your `loop` function.

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+  len <= String.length str &&
+  String.(sub str 0 len = pref)
+
+let is_app path =
+  let check appname =
+    let name = Filename.basename path in
+    Filename.check_suffix name ".app" &&
+    has_prefix name appname in
+  List.exists idas ~f:(fun ida ->
+      check (name64 ida) || check ida)
```
```suggestion
let is_ida_app path =
  List.exists ida_apps ~f:(Filename.check_suffix path)
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+      check (name64 ida) || check ida)
+
+let find_app path () =
+  match find path is_app with
```
```suggestion
  match find path is_ida_app with
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,232 @@
+  match find path is_app with
+  | None -> None
+  | Some path ->
+    let path' = path / "Contents/MacOS/" in
+    if Sys.file_exists path' then Some path'
```
```suggestion
    let path = path / "Contents/MacOS/" in
    if is_dir path then Some path
```

ivg reviewed (changes requested):
1. batch up my suggestions and commit them (then fix any errors I missed)
2. incorporate the `paths` function into the loop and count _all_ entries in the directory for the total maximum

gitoleg committed:
Apply suggestions from code review

Co-authored-by: Ivan Gotovchits <ivg@ieee.org>

gitoleg committed:
a couple minor bugs fixed

gitoleg committed:
uses Unix.readdir instead of Sys.readdir

gitoleg committed:
refactored

gitoleg committed:
refactored significantly

gitoleg committed:
debuggin

gitoleg committed:
debuggin

gitoleg committed:
debuggin

gitoleg committed:
debuggin

gitoleg committed:
everything works

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,250 @@
+  let rec read depth entries current dir =
+    if depth > max_depth || entries > max_paths
+    then leave dir None entries
+    else
+      match next dir with
```
```suggestion
    else match next dir with
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,250 @@
+        let path = current / entry in
+        if test path
+        then leave dir (Some path) entries
+        else
+          begin
+            match opendir path with
```
```suggestion
        else begin match opendir path with
```

ivg commented on packages/conf-ida/conf-ida.0.3/files/find-ida.ml.in:
```diff
@@ -0,0 +1,250 @@
+    let length = length t in
+    let rec loop i =
+      if i >= length then None
+      else
+        match f t.(i) with
```
```suggestion
      else match f t.(i) with
```

ivg reviewed (approved):
So, first of all, I really like the code, you caught the right idea of developing a DSL (a set of small functions) and expressing the logic using those small functions. 

I have several style suggestions, feel free to accept or reject them at your discretion.

gitoleg committed:
Apply suggestions from code review

Co-authored-by: Ivan Gotovchits <ivg@ieee.org>

ivg commented:
Besides, are CI failures at all relevant?

gitoleg commented:
well, `travis` still doesn't have any IDA installed, so no surprises tests fail

gitoleg committed:
constrains ocaml version
